jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/proxy/WebSocketProxy.java

```diff
@@ -0,0 +1,635 @@
+                        state = State.CONNECTING;
+                        client = session;
+                        break;
+
+                    default:
+                        failure = new IllegalStateException();
+                        break;
+                }
+            }
+
+            if (failure != null)
+                callback.failed(failure);
+            else
+                server2Proxy.connect(Callback.from(() -> onOpenSuccess(callback), (t) -> onOpenFail(callback, t)));
+        }
+
+        private void onOpenSuccess(Callback callback)
+        {
+            System.err.println(toString() + " onOpenSuccess()");
+
+            boolean failServer2Proxy = false;
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case CONNECTING:
+                        state = State.OPEN;
+                        break;
+
+                    case FAILED:
+                        failure = error;
+                        failServer2Proxy = true;
+                        break;
+
+                    default:
+                        failure = new IllegalStateException();
+                        break;
+                }
+            }
+
+            if (failServer2Proxy)
+                server2Proxy.fail(failure, callback);
+            else if (failure != null)
+                callback.failed(failure);
+            else
+                callback.succeeded();
+        }
+
+        private void onOpenFail(Callback callback, Throwable t)
+        {
+            System.err.println(toString() + " onOpenFail(): " + t);
+
+            Throwable failure = t;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case CONNECTING:
+                        state = State.FAILED;
+                        error = t;
+                        break;
+
+                    case FAILED:
+                        failure = error;
+                        break;
+
+                    default:
+                        failure = new IllegalStateException();
+                        break;
+                }
+            }
+
+            callback.failed(failure);
+        }
+
+        @Override
+        public void onFrame(Frame frame, Callback callback)
+        {
+            System.err.println(toString() + " onFrame(): " + frame);
+            receivedFrames.offer(Frame.copy(frame));
+
+            Callback sendCallback = callback;
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case OPEN:
+                        if (frame.getOpCode() == OpCode.CLOSE)
+                        {
+                            state = State.ICLOSED;
+                            closeCallback = callback;
+                            sendCallback = Callback.from(()->{}, callback::failed);
```

Comment:
You should add a comment about why you store `closeCallback` here - in 6 months from now you won't remember.
Note how the fact that you need to store the callback is because you _know_ what the callback is doing, and you want to delay that action.
This means that the javadocs for `onFrame()` should explain in great details what the callback is doing depending on the various frame types.
This is not ideal: you want different _methods_ to have different semantics, not the same method having different semantics based on the frame type.